### ISSUE TITLE
todo

### ISSUE THREAD

**Issue by lucidrains** [lucidrains/x-transformers-rl#1]
- [ ] world model
  - [x] fix caching during rollouts, keep track of sequence length in cache and set `seq_pos_start` (x-transformers work)
  - [ ] incorporate vision first with autoencoder, before moving onto other modalities
  - [ ] agent should be able to dream within world model
  - [x] schedule for world model being slowly incorporated into the actor / critic
  - [x] add hybrid GRU, but fix the inference caching for GRU in `x-transformers` first. look into gated delta net with state tracking patch
  - [ ] add local attention solution into x-transformers
  - [ ] allow for any multimodal encoding to and from world model
  - [x] use entropy of state prediction as exploration reward - in lieu of noisy nets
  - [ ] kl div loss w/ weight with reference world model
  - [ ] add state dropouts in the form of key padding attention mask during autoregressive training
- [ ] evolution
  - [ ] allow for core network to learn without latent gene conditioning, for research purposes
  - [ ] pass latent genes to world model with some dropout
  - [ ] think about how to incorporate occasional (1+1)evolutionary strategies, experiment with keeping track of gradient statistics to determine the proper noise std dev
  - [ ] allow for fusing the fittest latent / gene into the network, then renoising the latent mapper + repopulating for kickstarting exploration again. perhaps try this with island reset strategy at first
  - [ ] attempt to make one of the weights in actor / critic population based and figure out the best crossover strategy
  - [ ] start thinking about coevolution, with two population of genes, one for world model and the other for the actor / critic - for sleep phase learning
  - [ ] lamarckianism of genes w/ very small learning rate
  - [ ] think about how to involve map-elites for archiving, perhaps using technique from vq-map-elites paper from behavioral descriptors
  - [ ] naturally select the genes that gets to pass their gradients to the centralized world model, but start thinking about how world model can be broken into population based for coevo
  - [ ] interactable world model
  - [ ] in a population based setting, figure out some scheme to pass to world model the gradients from the fittest model, or gradients from individuals that agree with each other
  - [ ] bring in hypermutation scheme from AIS literature as an option in place of elitism
- [ ] actions
  - [x] discretize continuous
  - [x] multi discrete
  - [ ] hierarchical transformer for residual discretization
  - [ ] build out some action space module that can handle any combination of discretes, continuous, discretized continuous etc
- [ ] ppo
  - [ ] add post decision critic
  - [ ] make the actor / critic into feedforward stack + ada-ln zero conditioning with latent genes
  - [ ] allow for custom reward shaping functions
  - [ ] either have separate curiosity actor/critics, or have the existing ones accept some curiosity embed, with some prob for becoming playful during rollouts
  - [ ] look into [homeostatic rl](https://arxiv.org/abs/2205.08645)
- [ ] exploration
  - [ ] try the adversarially guided actor/critic paper
- [ ] simulators
  - [ ] mujoco
- [ ] continual learning
  - [ ] add shrink perturb scheduler into adam-atan2-pytorch, also add muon to adam-atan2 and make it customizable which layers and try out with the actor mlp head
- [ ] distillation related
  - [ ] grpo distillation from world model to actor
  - [ ] critic to actor with dual network, cite paper later
- [ ] memory
  - [ ] start with a simple fast weight matrix with delta updates
- [ ] engineering
  - [ ] make actor, critic, world model, more composable from the get go
